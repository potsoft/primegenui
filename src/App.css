.App {
  text-align: center;
}

.App-logo {
  height: 40vmin;
  pointer-events: none;
}

@media (prefers-reduced-motion: no-preference) {
  .App-logo {
    animation: App-logo-spin infinite 20s linear;
  }
}

.App-header {
  background-color: #282c34;
  min-height: 10vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  margin-bottom: 5vh;
  font-size: calc(10px + 2vmin);
  color: white;
}

.App-link {
  color: #61dafb;
}

.App-VSpacer {
  min-height: 5vh;
  display: flex;
  flex-direction: column;
  color: white;
}

.App-HSpacer {
  min-width: 2vw;
  display: flex;
  flex-direction: row;
  color: white;
}

.App-LimitEntry {
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: center;
  border: 0;
  color: #282c34;
  font-size: calc(5px + 2vmin);
}

.App-PrimesBox {
  display: flex;
  flex-direction: row;
  align-items: center;
  margin-top: 5vh;
  margin-right: 2vw;
  margin-bottom: 5vh;
  margin-left: 2vw;
  justify-content: center;
  border: 2px solid #282c34;
  color: #282c34;
  font-size: calc(5px + 2vmin);
}

.App-PrimesBox-Error-Headline {
  color: #ff0000;
}

.App-PrimesBox-OK-Headline {
  color: #282c34;
}


.App-PrimesWrap {
  display: flex;
  flex-wrap: wrap;
  align-items: center;
  justify-content: center;
}

input {
  padding: 10px;
  border-radius: 5px;
  margin-right: 10px;
  border: 3px solid #bbbbbb;
}

button {
  padding: 10px;
  border-radius: 5px;
  margin-right: 10px;
  border: 3px solid #bbbbbb;
  background: solid;
  background-color: #282c34;
  color: white;
  cursor: pointer;
}

