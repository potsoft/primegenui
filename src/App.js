import React, { Component } from 'react';
import './App.css';

// Constants in lieu of environment
const REST_API_PATH_HOST = 'http://localhost:9091';
const REST_API_PATH_PRIMES = '/primes';
const REST_API_PATH_PRIMES_LIMIT_PARAM = 'generationLimit';

// Other symbolic constants
const OK_MESSAGE_CSS = 'App-PrimesBox-OK-Headline';
const ERROR_MESSAGE_CSS = 'App-PrimesBox-Error-Headline';

/**
 * Simple, React-based UI for prime number generation example (Sieve of Erathosthenses).
 */
class App extends Component {
  constructor(props) {
    super(props);

    this.state = {
      // input entered by user that specifies maximum prime number to generate.
      generationLimit: '',
      // CSS class for message about success or failure of primes generation.
      messageCSS: OK_MESSAGE_CSS,
      // message about success or failure of primes generation, can be returned from primes service
      message: 'no primes generated yet!',
      // generated primes numbers, returned from primes service
      generatedPrimes: []
    };

    this.setGenerationLimit = this.setGenerationLimit.bind(this);
    this.generatePrimesIfValid = this.generatePrimesIfValid.bind(this);
  }

  render() {
    const primesTableCells = this.state.generatedPrimes.map((prime, idx) =>
      <p data-testid={`prime-nbr-${prime}`} key={prime}>{prime + (this.state.generatedPrimes.length === (idx+1) ? '' : ',')}</p>
    );
    const primesTableEmpty = <span className="App-VSpacer" data-testid="primesEmpty" />;

    return (
      <div className="App">

        <header className="App-header">
          <p>Prime Number Generation Example UI</p>
        </header>

        <form onSubmit={this.generatePrimesIfValid}>
          <fieldset className="App-LimitEntry">
            <legend>Generate Primes</legend>
            <label htmlFor="limit">From Zero To:</label>
            <span className="App-HSpacer" />
            <input name="limit" type="text" data-testid="limitInput" value={this.state.generationLimit} onChange={this.setGenerationLimit} />
            <span className="App-HSpacer" />
            <button type="submit">Go!</button>
          </fieldset>

        </form>

        <fieldset className="App-PrimesBox">
          <legend className={this.state.messageCSS}>{this.state.message}</legend>
          <div className="App-PrimesWrap">
            { this.state.generatedPrimes.length === 0 ? primesTableEmpty: primesTableCells }
          </div>
        </fieldset>

      </div>
    );
  }

  setGenerationLimit(event) {
    this.setState({ generationLimit: event.target.value });
    console.log('setGenerationLimit=' + this.state.generationLimit);
  }

  generatePrimesIfValid(event) {
    console.log('generatePrimesIfValid=' + this.state.generationLimit);
    event.preventDefault();

    // Basic client-side validation of generation limit entered by user; don't
    // call primes service if invalid.

    if (!this.validateGenerationLimit(this.state.generationLimit)) {
      this.setState({
        messageCSS: ERROR_MESSAGE_CSS,
        message: 'Entered generationLimit is not a valid integer',
        generatedPrimes: []
      });
      return;
    }

    // URL to primes service; no special HTTP headers in this example.

    let url = `${REST_API_PATH_HOST}${REST_API_PATH_PRIMES}?${REST_API_PATH_PRIMES_LIMIT_PARAM}=${this.state.generationLimit}`;
    console.log('generatePrimesIfValid(URL)=' + url);

    // Invokes the primes service synchronously in this example for expediency.

    fetch(url).then(response => {
      let httpStatus = response.status;
      this.setState({
        // change CSS of message to highlight errors detected by API.
        messageCSS: httpStatus === 200 ? OK_MESSAGE_CSS : ERROR_MESSAGE_CSS
      });
      return response.json();
    }).then(result => this.setState({
      // Set message and primes array from the response if API successfully invoked;
      // API may still have detected an error and returned HTTP status other than OK.
      message: result.message,
      generatedPrimes: result.generatedPrimes
    })).catch(error => this.setState({
      // If not API successfully invoked, clear primes array and set message to fetch error.
      messageCSS: ERROR_MESSAGE_CSS,
      message: `Primes service not available: ${error}`,
      generatedPrimes: []
    }));

    console.log('generatePrimesIfValid SUCCESS!');
  }

  /** Validate the primes generation limit as being an non-blank integer. */
  validateGenerationLimit(limit) {
    let isValid = false;
    if (limit && (typeof limit === 'string')) {
      const numFromStr = Number(limit);
      isValid = Number.isInteger(numFromStr);
    }
    return isValid;
  }
}

export default App;
