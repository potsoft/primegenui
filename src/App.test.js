import React from 'react';
import { render, screen, fireEvent, wait, waitFor } from '@testing-library/react';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom';
import App from './App';

beforeEach(() => {
  fetch.resetMocks();
});

/** Sanity check of initial rendering of app before any input entered. */
test('renders no-primes-generated-yet legend', () => {
  render(<App />);
  const legendElement = screen.getByText(/no primes generated yet/i);
  expect(legendElement).toBeInTheDocument();
});

/** Snap shot of initial rendering of app before any input entered. */
it('renders snapshot correctly when no-primes-generated-yet', () => {
  const tree = renderer.create(<App />).toJSON();
  expect(tree).toMatchSnapshot();
});

/** Mock successful call to primes service and verify generated prime number content. */
it('renders correctly when valid prime generation limit is provided (10)', async() => {
  fetch.mockResponseOnce(
    JSON.stringify({ message: "primes generated!", generatedPrimes: [2,3,5,7] }),
    { status: 200 });
  const { getByText, getByLabelText,getByPlaceholderText, getByTestId, container } = render(<App />);
  const input = getByTestId('limitInput');
  fireEvent.change(input, {target: {value: '10'}});
  fireEvent.click(getByText('Go!'));
  const elemWithPrime = await waitFor(() => getByTestId('prime-nbr-7'));
  expect(getByTestId('prime-nbr-7')).toHaveTextContent('7');
  expect(fetch).toHaveBeenCalledTimes(1);
});

/** Mock call to primes service returning error and verify no prime number content was generated. */
it('renders no primes when prime generation limit is invalid (zero)', async() => {
  fetch.mockResponseOnce(
    JSON.stringify({ message: "generationLimit cannot be zero or less than zero", generatedPrimes: [] }),
    { status: 422 });
  const { getByText, getByLabelText,getByPlaceholderText, getByTestId, container } = render(<App />);
  const input = getByTestId('limitInput');
  fireEvent.change(input, {target: {value: '0'}});
  fireEvent.click(getByText('Go!'));
  const elemEmpty = await waitFor(() => getByTestId('primesEmpty'));
  expect(getByTestId('primesEmpty')).toHaveTextContent('');
  expect(fetch).toHaveBeenCalledTimes(1);
})

/** Verify no prime number content was generated when client detectes entry of invalid limit. */
it('renders no primes when prime generation limit is invalid (non-numeric)', async() => {
  const { getByText, getByLabelText,getByPlaceholderText, getByTestId, container } = render(<App />);
  const input = getByTestId('limitInput');
  fireEvent.change(input, {target: {value: 'abc'}});
  fireEvent.click(getByText('Go!'));
  expect(getByTestId('primesEmpty')).toHaveTextContent('');
})


