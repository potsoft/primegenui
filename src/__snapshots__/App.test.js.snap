// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders snapshot correctly when no-primes-generated-yet 1`] = `
<div
  className="App"
>
  <header
    className="App-header"
  >
    <p>
      Prime Number Generation Example UI
    </p>
  </header>
  <form
    onSubmit={[Function]}
  >
    <fieldset
      className="App-LimitEntry"
    >
      <legend>
        Generate Primes
      </legend>
      <label
        htmlFor="limit"
      >
        From Zero To:
      </label>
      <span
        className="App-HSpacer"
      />
      <input
        data-testid="limitInput"
        name="limit"
        onChange={[Function]}
        type="text"
        value=""
      />
      <span
        className="App-HSpacer"
      />
      <button
        type="submit"
      >
        Go!
      </button>
    </fieldset>
  </form>
  <fieldset
    className="App-PrimesBox"
  >
    <legend
      className="App-PrimesBox-OK-Headline"
    >
      no primes generated yet!
    </legend>
    <div
      className="App-PrimesWrap"
    >
      <span
        className="App-VSpacer"
        data-testid="primesEmpty"
      />
    </div>
  </fieldset>
</div>
`;
